executor/join_result_generators.go

```diff
@@ -32,21 +32,51 @@ var (
 	_ joinResultGenerator = &innerJoinResultGenerator{}
 )
 
-// joinResultGenerator is used to generate join results according the join type, see every implementor for detailed information.
+// joinResultGenerator is used to generate join results according the join type.
+// A typical instruction flow is:
+//
+//     hasMatch := false
+//     for innerIter.Current() != innerIter.End() {
+//         matched, err := g.tryToMatch(outer, innerIter, chk)
+//         // handle err
+//         hasMatch = hasMatch || matched
+//     }
+//     if !hasMatch {
+//         g.onMissMatch(outer)
+//     }
+//
+// NOTE: This interface is **not** thread-safe.
 type joinResultGenerator interface {
-	// emit tries to join an outer row with a batch of inner rows.
-	// When inners == nil or inners.Len() == 0, it means that the outer row can not be joined with any inner row:
-	//     1. SemiJoin:              unmatched outer row is ignored.
-	//     2. AntiSemiJoin:          unmatched outer row is appended to the result buffer.
-	//     3. LeftOuterSemiJoin:     unmatched outer row is appended with 0 and appended to the result buffer.
-	//     4. AntiLeftOuterSemiJoin: unmatched outer row is appended with 1 and appended to the result buffer.
-	//     5. LeftOuterJoin:         unmatched outer row is joined with a row of NULLs and appended to the result buffer.
-	//     6. RightOuterJoin:        unmatched outer row is joined with a row of NULLs and appended to the result buffer.
-	//     7. InnerJoin:             unmatched outer row is ignored.
-	// When inners.Len != 0 but all the joined rows are filtered, this means that the outer row is unmatched and the above action is tacked as well.
-	// Otherwise, the outer row is matched and some joined rows is appended to the `chk`.
-	// The size of `chk` is MaxChunkSize at most.
-	emit(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) error
+	// tryToMatch tries to join an outer row with a batch of inner rows. When
+	// 'inners.Len != 0' but all the joined rows are filtered, the outer row is
+	// considered unmatched. Otherwise, the outer row is matched and some joined
+	// rows is appended to `chk`. The size of `chk` is limited to MaxChunkSize.
+	//
+	// NOTE: Callers need to call this function multiple times to consume all
+	// the inner rows for an outer row, and dicide whether the outer row can be
+	// matched with at lease one inner row.
+	tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (bool, error)
+
+	// onMissMatch operates on the unmatched outer row according to the join
+	// type. An outer row can be considered miss matched if:
+	//   1. it can not pass the filter on the outer table side.
+	//   2. there does not exist an inner row with the same join key.
```

Comment:
Maybe you can use "there is no inner row with the same join key."  Just for your reference.